### PR TITLE
Remove needless returns detected by clippy in the compiler

### PR DESCRIPTION
**Pull request by eduardosm**

### CODE REVIEW

**Comment by workingjubilee** [comment]
@eduardosm can you split this between compiler and library?

**Comment by eduardosm** [comment]
Done, libraries part [here](https://github.com/rust-lang/rust/pull/130115)

**Comment by RalfJung** [comment]
It's not obvious that these are all harmless changes - sometimes the explicit return expresses intent, so it should be preserved to account for future possible refactorings even if it does not make a difference right now.

**Comment by eduardosm** [comment]
I undid some of the changes, trying to keep only the most obvious ones.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 1]
```diff
@@ -336,7 +336,7 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 return true;
             }
         }
-        return false;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:339]: IMO it is nicer, for symmetry with the `return` just a few lines above, to use `return` here.

That's exactly why I don't like just forcing such clippy lints over the entire codebase...

Comment (saethlin) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:339]: I have exactly the opposite opinion. I like the consistency that tools like clippy encourage, and in this case I think the change makes the "early return" and "default/fallthrough" visually distinct, which I find easier to read.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:339]: The last two returns here are not distinct in a meaningful sense, so it is not a good idea to make them visually distinct.

**Comment by compiler-errors** [comment]
Oh cool, yay. Glad that's getting fixed.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs]: I agree with @saethlin. I think it's just plain odd to have `return value;` at the end of a block when it's not suggesting exceptional control flow.

I think it's actually very odd that miri has `allow(clippy::needless_return)` on its codebase, but for the *rest* of the compiler I think this should be fixed.

**Review by compiler-errors — APPROVED** [review]
I ended up implementing the exact same changes, so I basically did a complete review of this by doing the changes myself. This LGTM.

**Pull request merged**